arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -0,0 +1,160 @@
+    application_features::ApplicationServer& server)
+    : ApplicationFeature{server, "ClusterMetrics"} {
+  setOptional();
+  startsAfter<ClusterFeature>();
+  startsAfter<NetworkFeature>();
+  startsAfter<SchedulerFeature>();
+}
+
+void ClusterMetricsFeature::validateOptions(
+    std::shared_ptr<options::ProgramOptions> /*options*/) {
+  if (!ServerState::instance()->isCoordinator()) {
+    disable();
+  }
+}
+
+void ClusterMetricsFeature::asyncUpdate() {
+  if (isEnabled() && _count.fetch_add(1, std::memory_order_acq_rel) == 0) {
+    forceAsyncUpdate();
+  }
+}
+
+void ClusterMetricsFeature::forceAsyncUpdate() {
+  auto internal = [this] {
+    _count.store(1, std::memory_order_relaxed);
+    // TODO try get from agency
+    metricsOnCoordinator(server().getFeature<NetworkFeature>(),
+                         server().getFeature<ClusterFeature>())
+        .then([this](futures::Try<RawDBServers>&& metrics) mutable {
+          bool needRetry = !metrics.hasValue();
+          if (!needRetry) {
+            _lastUpdate = std::chrono::steady_clock::now();
+            // We want more time than kTimeout should expire after last cross
+            // cluster communication
+            set(std::move(*metrics));
+            needRetry = _count.fetch_sub(1, std::memory_order_acq_rel) != 1;
+            // if true then was call asyncUpdate() after _count.store(1)
+          }
+          if (needRetry) {
+            forceAsyncUpdate();
+          }
+        });  // detach() here
+  };
+  auto now = std::chrono::steady_clock::now();
+  if (now < _lastUpdate + kTimeout) {
+    _handle = SchedulerFeature::SCHEDULER->queueDelayed(
+        RequestLane::CLUSTER_INTERNAL, _lastUpdate + kTimeout - now,
+        [&](bool canceled) {
+          if (!canceled) {
+            internal();
+          }
+        });
+  } else {
+    _handle = nullptr;
+    SchedulerFeature::SCHEDULER->queue(RequestLane::CLUSTER_INTERNAL, internal);
+  }
+}
+
+void ClusterMetricsFeature::set(RawDBServers&& raw) const {
+  CoordinatorMetrics metrics;
+  {
+    std::shared_lock lock{_m};
+    for (auto const& payload : raw) {
+      TRI_ASSERT(payload);
+      velocypack::Slice slice{payload->data()};
+      size_t size = slice.length();
+      TRI_ASSERT(size % 3 == 0);
+      for (size_t i = 0; i != size; i += 3) {
+        auto name = slice.at(i).stringView();
+        auto labels = slice.at(i + 1);
+        auto value = slice.at(i + 2);
+        auto f = _toCoordinator.find(name);
+        if (f != _toCoordinator.end()) {
+          f->second(metrics, name, labels, value);
+        }
+      }
+    }
+  }
+  auto data = std::make_shared<Data>(std::move(metrics));
+  std::atomic_store_explicit(&_data, data, std::memory_order_release);
+  // TODO(MBkkt) serialize to velocypack array
+  // TODO(MBkkt) set to agency
+}
+
+void ClusterMetricsFeature::add(std::string_view metric,
+                                ToCoordinator const& toCoordinator) {
```

Comment:
Does it make sense to accept the `ToCoordinator` parameter by value here, and move it into `_toCoordinator[metric]` below?